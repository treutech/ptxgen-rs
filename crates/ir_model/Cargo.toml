
[package]
name = "ir-model"
version = "0.1.0"
edition = "2024"
license = "Apache-2.0"
authors = ["Raul Estrada <restrada@treutech.io>"]
homepage = "https://treutech.io/ptxgen-rs"
repository = "https://github.com/treutech/ptxgen-rs"

[dependencies]
serde = { version = "1.0.219", features = ["derive"] }
