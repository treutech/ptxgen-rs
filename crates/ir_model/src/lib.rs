// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use serde::Serialize;

#[derive(Debug, Serialize)]
pub enum Instruction {
    Load {
        function: String,
        dst: String,
        src: String,
    },
    Store {
        function: String,
        dst: String,
        value: String,
    },
    Add {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    FAdd {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    FMul {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    ICmp {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
        op: String,
    },
    GetElementPtr {
        function: String,
        dst: String,
        base: String,
        index: String,
    },
    Alloca {
        function: String,
        dst: String,
        ty: String,
        align: u32,
    },
    Br {
        function: String,
        cond: Option<String>,
        target_true: String,
        target_false: Option<String>,
    },
    Ret {
        function: String,
    },
    Unhandled {
        function: String,
        text: String,
    },
}

impl Instruction {
    pub fn function_name(&self) -> &str {
        match self {
            Instruction::Load { function, .. }
            | Instruction::Store { function, .. }
            | Instruction::Add { function, .. }
            | Instruction::FAdd { function, .. }
            | Instruction::FMul { function, .. }
            | Instruction::ICmp { function, .. }
            | Instruction::GetElementPtr { function, .. }
            | Instruction::Alloca { function, .. }
            | Instruction::Br { function, .. }
            | Instruction::Ret { function }
            | Instruction::Unhandled { function, .. } => function,
        }
    }

    pub fn used_operands(&self) -> Vec<&str> {
        use Instruction::*;
        match self {
            FMul { dst, lhs, rhs, .. }
            | FAdd { dst, lhs, rhs, .. }
            | Add { dst, lhs, rhs, .. }
            | ICmp { dst, lhs, rhs, .. } => vec![dst, lhs, rhs],

            Load { dst, src, .. } => vec![dst, src],
            Store { dst, value, .. } => vec![dst, value],

            Alloca { dst, .. } => vec![dst],
            GetElementPtr {
                dst, base, index, ..
            } => vec![dst, base, index],

            Unhandled { text, .. } => vec![text],
            _ => vec![],
        }
    }
}
