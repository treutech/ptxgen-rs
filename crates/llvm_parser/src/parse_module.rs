use anyhow::{Context, Result};
use llvm_ir::Module;
use std::fs;
use std::path::Path;

/// Parse an LLVM IR (.ll) file into an `llvm_ir::Module`
pub fn parse_module<P: AsRef<Path>>(path: P) -> Result<Module> {
    let ll_text = fs::read_to_string(&path)
        .with_context(|| format!("Failed to read file: {}", path.as_ref().display()))?;

    Module::from_ir_str(&ll_text)
        .map_err(|e| anyhow::anyhow!(e))
        .with_context(|| format!("Failed to parse LLVM IR in file: {}", path.as_ref().display()))
}
