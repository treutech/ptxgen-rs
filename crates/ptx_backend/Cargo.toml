
[package]
name = "ptx-backend"
version = "0.1.0"
edition = "2024"
license = "Apache-2.0"
authors = ["Raul Estrada <restrada@treutech.io>"]
homepage = "https://treutech.io/ptxgen-rs"
repository = "https://github.com/treutech/ptxgen-rs"

[dependencies]
ir-model = { path = "../ir_model" }
llvm-parser = { path = "../llvm_parser" }
clap = { version = "4.5.41", features = ["derive"] }
assert_cmd = "2.0.17"
anyhow = "1.0.98"
llvm-ir = "0.11.3"

[[bin]]
name = "ptx-backend"
path = "src/main.rs"

[dev-dependencies]
insta = "1.43.1"
