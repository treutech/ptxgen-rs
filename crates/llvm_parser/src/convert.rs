// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use ir_model::Instruction;
use llvm_ir::instruction::Instruction as LlvmInst;

pub fn lower(function: &str, instr: &LlvmInst) -> Instruction {
    use LlvmInst::*;
    match instr {
        FMul(f) => Instruction::FMul {
            function: function.to_string(),
            dst: f.dest.to_string(),
            lhs: f.operand0.to_string(),
            rhs: f.operand1.to_string(),
        },
        FAdd(f) => Instruction::FAdd {
            function: function.to_string(),
            dst: f.dest.to_string(),
            lhs: f.operand0.to_string(),
            rhs: f.operand1.to_string(),
        },
        Load(l) => Instruction::Load {
            function: function.to_string(),
            dst: l.dest.to_string(),
            src: l.address.to_string(),
        },
        Store(s) => Instruction::Store {
            function: function.to_string(),
            dst: s.address.to_string(),
            value: s.value.to_string(),
        },
        Alloca(a) => Instruction::Alloca {
            function: function.to_string(),
            dst: a.dest.to_string(),
            ty: format!("{:?}", a.allocated_type),
            align: a.alignment,
        },
        ICmp(cmp) => Instruction::ICmp {
            function: function.to_string(),
            dst: cmp.dest.to_string(),
            op: format!("{:?}", cmp.predicate),
            lhs: cmp.operand0.to_string(),
            rhs: cmp.operand1.to_string(),
        },
        FCmp(cmp) => Instruction::FCmp {
            function: function.to_string(),
            dst: cmp.dest.to_string(),
            op: format!("{:?}", cmp.predicate),
            lhs: cmp.operand0.to_string(),
            rhs: cmp.operand1.to_string(),
        },
        Add(add) => Instruction::Add {
            function: function.to_string(),
            dst: add.dest.to_string(),
            lhs: add.operand0.to_string(),
            rhs: add.operand1.to_string(),
        },
        Sub(s) => Instruction::Sub {
            function: function.to_string(),
            dst: s.dest.to_string(),
            lhs: s.operand0.to_string(),
            rhs: s.operand1.to_string(),
        },
        FSub(s) => Instruction::FSub {
            function: function.to_string(),
            dst: s.dest.to_string(),
            lhs: s.operand0.to_string(),
            rhs: s.operand1.to_string(),
        },
        Mul(m) => Instruction::Mul {
            function: function.to_string(),
            dst: m.dest.to_string(),
            lhs: m.operand0.to_string(),
            rhs: m.operand1.to_string(),
        },
        UDiv(d) => Instruction::UDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        SDiv(d) => Instruction::SDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        URem(r) => Instruction::URem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        SRem(r) => Instruction::SRem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        FDiv(d) => Instruction::FDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        FRem(r) => Instruction::FRem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        GetElementPtr(gep) => Instruction::GetElementPtr {
            function: function.to_string(),
            dst: gep.dest.to_string(),
            base: gep.address.to_string(),
            index: gep
                .indices
                .iter()
                .map(|i| i.to_string())
                .collect::<Vec<_>>()
                .join(", "),
        },
        Phi(p) => Instruction::Phi {
            function: function.to_string(),
            dst: p.dest.to_string(),
            incoming: p
                .incoming_values
                .iter()
                .map(|(val, label)| {
                    (
                        match label {
                            llvm_ir::Name::Name(s) => s.as_str().to_string(),
                            llvm_ir::Name::Number(n) => format!("{}", n),
                        },
                        val.to_string(),
                    )
                })
                .collect(),
        },
        Select(sel) => Instruction::Select {
            function: function.to_string(),
            dst: sel.dest.to_string(),
            cond: sel.condition.to_string(),
            val_true: sel.true_value.to_string(),
            val_false: sel.false_value.to_string(),
        },
        BitCast(bc) => Instruction::Bitcast {
            function: function.to_string(),
            dst: bc.dest.to_string(),
            src: bc.operand.to_string(),
        },
        ZExt(z) => Instruction::ZExt {
            function: function.to_string(),
            dst: z.dest.to_string(),
            src: z.operand.to_string(),
        },
        Trunc(t) => Instruction::Trunc {
            function: function.to_string(),
            dst: t.dest.to_string(),
            src: t.operand.to_string(),
        },
        Call(c) => {
            let target = match &c.function {
                either::Either::Right(llvm_ir::Operand::ConstantOperand(const_ref)) => {
                    match const_ref.as_ref() {
                        llvm_ir::constant::Constant::GlobalReference { name, .. } => match name {
                            llvm_ir::Name::Name(s) => s.to_string(),
                            llvm_ir::Name::Number(n) => format!("{}", n),
                        },
                        _ => "unknown_fn".to_string(),
                    }
                }
                _ => "unknown_fn".to_string(),
            };

            Instruction::Call {
                function: function.to_string(),
                target,
                args: c.arguments.iter().map(|a| a.0.to_string()).collect(),
            }
        }
        _ => Instruction::Unhandled {
            function: function.to_string(),
            text: format!("{:?}", instr),
        },
    }
}

use llvm_ir::Terminator;

pub fn lower_terminator(func: &str, term: &Terminator) -> Instruction {
    match term {
        Terminator::Ret(_) => Instruction::Ret {
            function: func.to_string(),
        },
        Terminator::Br(br) => Instruction::Br {
            function: func.to_string(),
            cond: None,
            target_true: br.dest.to_string(),
            target_false: None,
        },
        Terminator::CondBr(br) => Instruction::Br {
            function: func.to_string(),
            cond: Some(format!("{}", br.condition)),
            target_true: br.true_dest.to_string(),
            target_false: Some(br.false_dest.to_string()),
        },
        _ => Instruction::Unhandled {
            function: func.to_string(),
            text: format!("{:?}", term),
        },
    }
}
