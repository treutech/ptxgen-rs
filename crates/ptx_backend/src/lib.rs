// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub mod ptx_type;
pub mod utils;
pub mod type_map;

use crate::ptx_type::PTXType;
use crate::utils::{clean_operand, get_register_type};
use ir_model::Instruction;
use crate::type_map::{TypeMap, declare_registers_from_typemap};


pub fn lower_function(
    name: &str,
    all_instrs: &[(String, Vec<Instruction>)],
    target: &str,
) -> Vec<String> {
    let mut output = vec![];

    output.push(format!("// Function: {}", name));
    output.push(emit_header(target));
    output.push(format!(".entry {} {{", clean_operand(name)));

    let flat_instrs: Vec<&Instruction> = all_instrs
        .iter()
        .flat_map(|(_, instrs)| instrs.iter())
        .collect();

    let mut type_map = TypeMap::new();
    for instr in &flat_instrs {
        for operand in instr.used_operands() {
            if let Some(ty_str) = get_register_type(instr, &operand) {
                let ptx_ty = PTXType::from_str(ty_str);
                type_map.insert(&clean_operand(&operand), ptx_ty);
            }
        }
    }
    for line in declare_registers_from_typemap(&type_map) {
        output.push(line);
    }

    for (block_name, instrs) in all_instrs {
        if instrs.is_empty() {
            continue;
        }
        output.push(format!("{}:", clean_operand(block_name)));
        for instr in instrs {
            let line = to_ptx(instr, &type_map);
            output.push(format!("    {}", line));
        }
    }

    let last_instr = flat_instrs.iter().rev().find(|i| {
        !matches!(
            i,
            Instruction::Alloca { .. } | Instruction::GetElementPtr { .. }
        )
    });

    if let Some(instr) = last_instr {
        match instr {
            Instruction::Ret { .. } | Instruction::Br { .. } => {}
            _ => output.push("    ret;".into()),
        }
    } else {
        output.push("    ret;".into());
    }

    output.push("}".into());
    output
}

fn emit_header(target: &str) -> String {
    format!(".version 7.0\n.target {}\n.address_size 64\n", target)
}

pub fn to_ptx(instr: &Instruction, type_map: &TypeMap) -> String {
    use Instruction::*;

    fn reg(op: &str) -> String {
        let clean = clean_operand(op);
        if clean.starts_with('%') {
            clean
        } else {
            format!("%{}", clean)
        }
    }

    fn mem(op: &str) -> String {
        let clean = clean_operand(op);
        if clean.starts_with('%') {
            format!("[{}]", clean)
        } else {
            format!("[%{}]", clean)
        }
    }

    fn extract_arg_name(arg: &str) -> String {
        if arg.starts_with("i32 ") || arg.starts_with("f32 ") {
            arg.split_whitespace().last().unwrap_or(arg).to_string()
        } else {
            arg.to_string()
        }
    }

    match instr {
        FMul { dst, lhs, rhs, .. } => {
            format!("mul.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FAdd { dst, lhs, rhs, .. } => {
            format!("add.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Add { dst, lhs, rhs, .. } => {
            format!("add.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Sub { dst, lhs, rhs, .. } => {
            format!("sub.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FSub { dst, lhs, rhs, .. } => {
            format!("sub.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Mul { dst, lhs, rhs, .. } => {
            format!("mul.lo.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        UDiv { dst, lhs, rhs, .. } => {
            format!("div.u32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        SDiv { dst, lhs, rhs, .. } => {
            format!("div.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        URem { dst, lhs, rhs, .. } => {
            format!("rem.u32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        SRem { dst, lhs, rhs, .. } => {
            format!("rem.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FDiv { dst, lhs, rhs, .. } => {
            format!("div.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FRem { dst, lhs, rhs, .. } => {
            format!("rem.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        ICmp {
            dst, lhs, rhs, op, ..
        } => {
            let pred = match op.as_str() {
                "EQ" => "eq",
                "NE" => "ne",
                "UGT" | "SGT" => "gt",
                "UGE" | "SGE" => "ge",
                "ULT" | "SLT" => "lt",
                "ULE" | "SLE" => "le",
                _ => {
                    return format!("// unsupported icmp predicate: {}", op);
                }
            };
            format!(
                "setp.{}.s32 {}, {}, {};",
                pred,
                reg(dst),
                reg(lhs),
                reg(rhs)
            )
        }
        FCmp {
            dst, lhs, rhs, op, ..
        } => {
            let pred = match op.as_str() {
                "OEQ" | "UEQ" => "eq",
                "ONE" | "UNE" => "ne",
                "OGT" | "UGT" => "gt",
                "OGE" | "UGE" => "ge",
                "OLT" | "ULT" => "lt",
                "OLE" | "ULE" => "le",
                _ => "lt", // fallback
            };
            format!(
                "setp.{}.f32 {}, {}, {};",
                pred,
                reg(dst),
                reg(lhs),
                reg(rhs)
            )
        }
        Load { dst, src, .. } => {
            let ty = type_map
                .get(&clean_operand(dst))
                .unwrap_or(&PTXType::S32)
                .as_str();

            let space = "global";
            format!("ld.{space}.{ty} {}, [{}];", reg(dst), clean_operand(src))
        }
        Store { dst, value, .. } => {
            let ty = type_map
                .get(&clean_operand(value))
                .unwrap_or(&PTXType::S32)
                .as_str();

            let space = "global";
            format!("st.{space}.{ty} {}, {};", mem(dst), reg(value))
        }
        Br {
            cond,
            target_true,
            target_false,
            ..
        } => match (cond, target_false) {
            (Some(c), Some(f)) => format!("@{} bra {};\n    bra {};", reg(c), target_true, f),
            (None, _) => format!("bra {};", target_true),
            _ => "// invalid conditional branch".to_string(),
        },
        CondBr {
            cond,
            then_target,
            else_target,
            ..
        } => {
            format!(
                "@{cond} bra {then};\n    bra {els};",
                cond = reg(cond),
                then = then_target,
                els = else_target
            )
        }
        Ret { .. } => "ret;".to_string(),
        GetElementPtr {
            dst, base, index, ..
        } => {
            let base_clean = clean_operand(base);
            let dst_clean = clean_operand(dst);
            let offset = format!("{}_offset", dst_clean);
            let calc_offset = format!("mul.lo.s32 %{}, %{}, 4;", offset, clean_operand(index));
            let calc_ptr = format!("add.s32 %{}, %{}, %{};", dst_clean, base_clean, offset);
            format!("{calc_offset}\n    {calc_ptr}")
        }
        Phi { dst, incoming, .. } => {
            let ty = type_map
                .get(&clean_operand(dst))
                .unwrap_or(&PTXType::S32)
                .as_str();
            format!("// phi.{ty} {} <- {:?}", reg(dst), incoming)
        }
        Alloca { .. } => String::new(),
        Select {
            dst,
            cond,
            val_true,
            val_false,
            ..
        } => {
            let ty = type_map
                .get(&clean_operand(dst))
                .unwrap_or(&PTXType::S32)
                .as_str();
            format!(
                "selp.{ty} {}, {}, {}, {};",
                reg(dst),
                reg(val_true),
                reg(val_false),
                reg(cond)
            )
        }
        Bitcast { dst, src, .. } => {
            format!("mov.b32 {}, {};", reg(dst), reg(src))
        }
        ZExt { dst, src, .. } => {
            format!("cvt.u32.u8 {}, {};", reg(dst), reg(src))
        }
        Trunc { dst, src, .. } => {
            format!("cvt.u8.u32 {}, {};", reg(dst), reg(src))
        }
        Call {
            callee, args, ret, ..
        } => {
            let mut ptx = String::new();

            if let Some(retvar) = ret {
                let ty = type_map
                    .get(&clean_operand(retvar))
                    .unwrap_or(&PTXType::S32)
                    .as_str();
                ptx.push_str(&format!("\t.param .{ty} retval_{};\n", retvar));
            }

            for (i, arg) in args.iter().enumerate() {
                let arg_val = extract_arg_name(arg);
                let ty = type_map
                    .get(&clean_operand(&arg_val))
                    .unwrap_or(&PTXType::S32)
                    .as_str();
                ptx.push_str(&format!("\t.param .{ty} arg{i};\n"));
            }

            for (i, arg) in args.iter().enumerate() {
                let arg_val = extract_arg_name(arg);
                let reg_name = reg(&arg_val);
                let ty = type_map
                    .get(&clean_operand(&arg_val))
                    .unwrap_or(&PTXType::S32)
                    .as_str();

                ptx.push_str(&format!("\tst.param.{ty} [arg{i}], {reg_name};\n"));
            }

            let arg_params = (0..args.len())
                .map(|i| format!("arg{i}"))
                .collect::<Vec<_>>()
                .join(", ");

            if let Some(retvar) = ret {
                ptx.push_str(&format!(
                    "\tcall (retval_{retvar}) {}, ({});\n",
                    clean_operand(callee),
                    arg_params
                ));
                let ty = type_map
                    .get(&clean_operand(retvar))
                    .unwrap_or(&PTXType::S32)
                    .as_str();
                ptx.push_str(&format!(
                    "\tld.param.{ty} {}, [retval_{}];\n",
                    reg(retvar),
                    retvar
                ));
            } else {
                ptx.push_str(&format!(
                    "\tcall {}, ({});\n",
                    clean_operand(callee),
                    arg_params
                ));
            }

            ptx
        }

        Unhandled { text, .. } => format!("// unhandled: {}", text),
    }
}

use anyhow::Result;
use llvm_ir::Module;
use llvm_parser::parse_llvm_ir_from_str;

pub fn compile_llvm_to_ptx(ir_code: &str) -> Result<String> {
    let module: Module = parse_llvm_ir_from_str(ir_code)?;
    let mut ptx_lines = vec![];

    for func in &module.functions {
        let blocks = llvm_parser::lower(func)?;
        let kernel_name = &func.name;

        // Count no-supported instructions per function
        let flat_instrs: Vec<&Instruction> = blocks
            .iter()
            .flat_map(|(_, instrs)| instrs.iter())
            .collect();

        let unhandled_count = flat_instrs
            .iter()
            .filter(|i| matches!(i, Instruction::Unhandled { .. }))
            .count();

        if unhandled_count > 0 {
            eprintln!(
                "Warning: {unhandled_count} unhandled instruction(s) in function `{}`",
                kernel_name
            );
        }

        let func_lines = lower_function(kernel_name, &blocks, "sm_75");
        ptx_lines.extend(func_lines);
        ptx_lines.push(String::new());
    }

    Ok(ptx_lines.join("\n"))
}
