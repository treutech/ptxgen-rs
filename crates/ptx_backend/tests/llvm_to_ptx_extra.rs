// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use llvm_parser::parse_module::parse_module;
use ir_model::Instruction;

fn run_test(filename: &str) -> String {
    use llvm_parser::convert::lower;
    use ptx_backend::lower_function;

    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("examples")
        .join(filename);
    let module = parse_module(&path).expect("Failed to parse module");

    let mut output = String::new();

    for func in module.functions {
        let all_instrs: Vec<(String, Vec<Instruction>)> = func
            .basic_blocks
            .into_iter()
            .map(|block| {
                let instrs = block.instrs.iter().map(lower).collect::<Vec<_>>();
                (block.name.to_string(), instrs)
            })
            .collect();
        
        for line in lower_function(&func.name, &all_instrs, "sm_75") {
            output.push_str(&format!("{}\n", line));
        }
    }

    output
}

#[test]
fn test_ret_only() {
    insta::assert_snapshot!("ret_only", run_test("ret_only.ll"));
}

#[test]
fn test_br_uncond() {
    insta::assert_snapshot!("br_uncond", run_test("br_uncond.ll"));
}

#[test]
fn test_br_cond() {
    insta::assert_snapshot!("br_cond", run_test("br_cond.ll"));
}

#[test]
fn test_phi_basic() {
    insta::assert_snapshot!("phi_basic", run_test("phi_basic.ll"));
}

#[test]
fn test_conflict_type() {
    insta::assert_snapshot!("conflict_type", run_test("conflict_type.ll"));
}
