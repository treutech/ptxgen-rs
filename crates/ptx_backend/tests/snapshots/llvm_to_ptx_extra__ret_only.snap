---
source: crates/ptx_backend/tests/llvm_to_ptx_extra.rs
expression: "run_test(\"ret_only.ll\")"
---
// Function: ret_only
.version 7.0
.target sm_75
.address_size 64

.entry ret_only {
    ret;
}
